server/src/main/java/org/opensearch/index/IndexService.java

```diff
@@ -525,7 +527,8 @@ public synchronized IndexShard createShard(
                 indexingOperationListeners,
                 () -> globalCheckpointSyncer.accept(shardId),
                 retentionLeaseSyncer,
-                circuitBreakerService
+                circuitBreakerService,
+                checkpointPublisher
```

Comment:
It doesn't look like we are conditionally invoking the new constructor here.  I think thats ok, but If we don't use it we shouldn't declare it in IndexShard.  We would then conditionally either wire up a real SegmentReplicationCheckpointPublisher in IndicesModule or `SegmentReplicationCheckpointPublisher.EMPTY`, depending on the feature flag from IndicesModule, and pass it down.

Alternatively, we could perhaps plug this in with a new constructor & method in EngineConfigFactory and avoid changing IndexShard's constructor entirely?  I think I would prefer that if it works because it is the engine that is being configured with the listener, not the IndexShard. Inside IndexShard's `newEngineConfig` method we could invoke a method on the engineConfigFactory to fetch a reference to the publisher, wire up the listener and pass it along.  This would extend nicely in the future if we want to plug in different strategies for where to publish checkpoints as a plugin.

Reply:
@nknize could you please provide your opinion here. We are not sure how to wire up new checkointPublisher, If we want to plug this in EngineConfigFactory, we have to pass it as plugins and it is making more complex. What are your thoughts on cuurent PR's way of interacting with IndexShard, would you like current logic to decouple from IndexShard and may be present in a different place?

Reply:
Maybe @Bukhtawar can comment here too?

Reply:
Took a closer look at this.  I think what I was suggesting isn't possible with where EngineConfigFactory is built in IndicesService.  We would have to create a new extension point within EnginePlugin.  Given we only currently have a single method of publishing checkpoints, I don't think we should overcomplicate this with that change.

Reply:
I think with `CheckpointRefreshListener ` being a part of the `IndexShard`, it's essentially the index shard that is being notified of a refresh in the engine, which is then taking care of publishing the checkpoint to the replica. Then the checkpoint request handler is also sitting on the corresponding replica `IndexShard` and synchronising the segment state. 
The current implementation looks fine to me based on the abstraction level IndexShard has

Reply:
Agree.  I was thinking of a way for us to construct and inject a list of listeners on the engine through config without having to pass this through IndexShard, but I don't think thats possible.